core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/internal/signer/V4RequestSigner.java

```diff
@@ -77,6 +80,8 @@ static V4RequestSigner query(V4Properties properties) {
                 requestBuilder.putRawQueryParameter(SignerConstant.X_AMZ_SECURITY_TOKEN,
                                                     ((AwsSessionCredentialsIdentity) properties.getCredentials()).sessionToken());
             }
+            // We have to add the host-header here explicitly, since query-signed request requires it in the signed-header param
+            addHostHeader(requestBuilder);
```

Comment:
I see addHostHeader logic here in query and presign, and in DefaultV4RequestSigner for header. If host header is needed in all cases, why duplicate the logic v/s move it higher up in single place?

Reply:
We can move it out of `DefaultV4RequestSigner`, sure.

Reply:
Out of `DefaultV4RequestSigner` and all in the static methods of V4RequestSigner is better than split across both sides. But it's still there multiple times. When I was thinking "move it higher up in single place", I was considering it in DefaultAwsV4HttpSigner before the requestSigner is invoked - did you mean that such manipulation of the requestBuilder to be responsibility of RequestSigner and have DefaultAwsV4HttpSigner just handle the logic flow between these different "processors" - request/payload/checksum.

Reply:
Yes, that was the thought. DefaultAwsV4HttpSigner should try to be focused on delegating rather than doing things itself.